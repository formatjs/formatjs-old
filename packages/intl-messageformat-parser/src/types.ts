export enum TYPE {
  /**
   * Raw text
   */
  literal,
  /**
   * Variable w/o any format, e.g `var` in `this is a {var}`
   */
  argument,
  /**
   * Variable w/ number format
   */
  number,
  /**
   * Variable w/ date format
   */
  date,
  /**
   * Variable w/ time format
   */
  time,
  /**
   * Variable w/ select format
   */
  select,
  /**
   * Variable w/ plural format
   */
  plural,
  /**
   * Only possible within plural argument.
   * This is the `#` symbol that will be substituted with the count.
   */
  pound,
}

export const enum SKELETON_TYPE {
  number,
  dateTime,
}

export interface LocationDetails {
  offset: number;
  line: number;
  column: number;
}
export interface Location {
  start: LocationDetails;
  end: LocationDetails;
}

export interface BaseElement<T extends TYPE> {
  type: T;
  value: string;
  location?: Location;
}

export type LiteralElement = BaseElement<TYPE.literal>;
export type ArgumentElement = BaseElement<TYPE.argument>;

export interface SimpleFormatElement<T extends TYPE, S extends Skeleton>
  extends BaseElement<T> {
  style?: string | S | null;
}

export type NumberElement = SimpleFormatElement<TYPE.number, NumberSkeleton>;
export type DateElement = SimpleFormatElement<TYPE.date, DateTimeSkeleton>;
export type TimeElement = SimpleFormatElement<TYPE.time, DateTimeSkeleton>;

export interface SelectOption {
  id: string;
  value: MessageFormatElement[];
  location?: Location;
}

export type ValidPluralRule =
  | 'zero'
  | 'one'
  | 'two'
  | 'few'
  | 'many'
  | 'other'
  | string;

export interface PluralOrSelectOption {
  value: MessageFormatElement[];
  location?: Location;
}

export interface SelectElement extends BaseElement<TYPE.select> {
  options: Record<string, PluralOrSelectOption>;
}

export interface PluralElement extends BaseElement<TYPE.plural> {
  options: Record<ValidPluralRule, PluralOrSelectOption>;
  offset: number;
  pluralType: Intl.PluralRulesOptions['type'];
}

export interface PoundElement {
  type: TYPE.pound;
  location?: Location;
}

export type MessageFormatElement =
  | LiteralElement
  | ArgumentElement
  | NumberElement
  | DateElement
  | TimeElement
  | SelectElement
  | PluralElement
  | PoundElement;

export interface NumberSkeletonToken {
  stem: string;
  options: string[];
}

export interface NumberSkeleton {
  type: SKELETON_TYPE.number;
  tokens: NumberSkeletonToken[];
  location?: Location;
}

export interface DateTimeSkeleton {
  type: SKELETON_TYPE.dateTime;
  pattern: string;
  location?: Location;
}

export type Skeleton = NumberSkeleton | DateTimeSkeleton;

/**
 * Type Guards
 */
export function isLiteralElement(
  el: MessageFormatElement
): el is LiteralElement {
  return el.type === TYPE.literal;
}
export function isArgumentElement(
  el: MessageFormatElement
): el is ArgumentElement {
  return el.type === TYPE.argument;
}
export function isNumberElement(el: MessageFormatElement): el is NumberElement {
  return el.type === TYPE.number;
}
export function isDateElement(el: MessageFormatElement): el is DateElement {
  return el.type === TYPE.date;
}
export function isTimeElement(el: MessageFormatElement): el is TimeElement {
  return el.type === TYPE.time;
}
export function isSelectElement(el: MessageFormatElement): el is SelectElement {
  return el.type === TYPE.select;
}
export function isPluralElement(el: MessageFormatElement): el is PluralElement {
  return el.type === TYPE.plural;
}
export function isPoundElement(el: MessageFormatElement): el is PoundElement {
  return el.type === TYPE.pound;
}
export function isNumberSkeleton(
  el: NumberElement['style'] | Skeleton
): el is NumberSkeleton {
  return !!(el && typeof el === 'object' && el.type === SKELETON_TYPE.number);
}
export function isDateTimeSkeleton(
  el?: DateElement['style'] | TimeElement['style'] | Skeleton
): el is DateTimeSkeleton {
  return !!(el && typeof el === 'object' && el.type === SKELETON_TYPE.dateTime);
}

export function createLiteralElement(value: string): LiteralElement {
  return {
    type: TYPE.literal,
    value,
  };
}

export function createNumberElement(
  value: string,
  style?: string | null
): NumberElement {
  return {
    type: TYPE.number,
    value,
    style,
  };
}

export interface Options {
  /**
   * Whether to convert `#` in plural rule options
   * to `{var, number}`
   * Default is true
   */
  normalizeHashtagInPlural?: boolean;
  /**
   * Capture location info in AST
   * Default is false
   */
  captureLocation?: boolean;
}
