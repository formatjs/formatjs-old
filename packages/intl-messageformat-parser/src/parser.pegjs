/*
Copyright 2014, Yahoo! Inc. All rights reserved.
Copyrights licensed under the New BSD License.
See the accompanying LICENSE file for terms.
*/

/*
Inspired by and derivied from:
messageformat.js https://github.com/SlexAxton/messageformat.js
Copyright 2014 Alex Sexton
Apache License, Version 2.0
*/

start
    = messageFormatPattern

messageFormatPattern
    = elements:messageFormatElement* {
        return {
            type    : 'messageFormatPattern',
            elements: elements
        };
    }

messageFormatElement
    = messageTextElement
    / argumentElement

messageText
    = text:(_ chars _)+ {
        var string = '',
            i, j, outerLen, inner, innerLen;

        for (i = 0, outerLen = text.length; i < outerLen; i += 1) {
            inner = text[i];

            for (j = 0, innerLen = inner.length; j < innerLen; j += 1) {
                string += inner[j];
            }
        }

        return string;
    }
    / $(ws)

messageTextElement
    = messageText:messageText {
        return {
            type : 'messageTextElement',
            value: messageText
        };
    }

argument
    = number
    / $([^ \t\n\r,.+={}#]+)

argumentElement
    = '{' _ id:argument _ format:(',' _ elementFormat)? _ '}' {
        return {
            type  : 'argumentElement',
            id    : id,
            format: format && format[2]
        };
    }

elementFormat
    = simpleFormat
    / pluralFormat
    / selectOrdinalFormat
    / selectFormat

simpleFormat
    = type:('number' / 'date' / 'time') _ style:(',' _ chars)? {
        return {
            type : type + 'Format',
            style: style && style[2]
        };
    }

pluralFormat
    = 'plural' _ ',' _ pluralStyle:pluralStyle {
        return {
            type   : pluralStyle.type,
            ordinal: false,
            offset : pluralStyle.offset || 0,
            options: pluralStyle.options
        };
    }

selectOrdinalFormat
    = 'selectordinal' _ ',' _ pluralStyle:pluralStyle {
        return {
            type   : pluralStyle.type,
            ordinal: true,
            offset : pluralStyle.offset || 0,
            options: pluralStyle.options
        }
    }

selectFormat
    = 'select' _ ',' _ options:optionalFormatPattern+ {
        return {
            type   : 'selectFormat',
            options: options
        };
    }

selector
    = $('=' number)
    / chars

optionalFormatPattern
    = _ selector:selector _ '{' _ pattern:messageFormatPattern _ '}' {
        return {
            type    : 'optionalFormatPattern',
            selector: selector,
            value   : pattern
        };
    }

offset
    = 'offset:' _ number:number {
        return number;
    }

pluralStyle
    = offset:offset? _ options:optionalFormatPattern+ {
        return {
            type   : 'pluralFormat',
            offset : offset,
            options: options
        };
    }

// -- Helpers ------------------------------------------------------------------

ws 'whitespace' = [ \t\n\r]+
_ 'optionalWhitespace' = $(ws*)

digit    = [0-9]
hexDigit = [0-9a-f]i

number = digits:('0' / $([1-9] digit*)) {
    return parseInt(digits, 10);
}

char
    = [^{}\\\0-\x1F\x7f \t\n\r]
    / '\\\\' { return '\\'; }
    / '\\#'  { return '\\#'; }
    / '\\{'  { return '\u007B'; }
    / '\\}'  { return '\u007D'; }
    / '\\u'  digits:$(hexDigit hexDigit hexDigit hexDigit) {
        return String.fromCharCode(parseInt(digits, 16));
    }

chars = chars:char+ { return chars.join(''); }
