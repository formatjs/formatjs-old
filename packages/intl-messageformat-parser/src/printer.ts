/*
 * Copyright 2015, Yahoo Inc.
 * Copyrights licensed under the New BSD License.
 * See the accompanying LICENSE file for terms.
 */

import {
  MessageFormatElement,
  isLiteralElement,
  LiteralElement,
  ArgumentElement,
  DateElement,
  TimeElement,
  NumberElement,
  SelectElement,
  PluralElement,
  isSelectElement,
  isArgumentElement,
  isDateElement,
  isTimeElement,
  isNumberElement,
  isPluralElement,
  TYPE,
  Skeleton,
  SKELETON_TYPE,
  NumberSkeletonToken,
  DateTimeSkeleton,
  isPoundElement,
} from './types';

export function printAST(ast: MessageFormatElement[]): string {
  return doPrintAST(ast, false);
}

export function doPrintAST(
  ast: MessageFormatElement[],
  isInPlural: boolean
): string {
  const printedNodes = ast.map(el => {
    if (isLiteralElement(el)) {
      return printLiteralElement(el, isInPlural);
    }

    if (isArgumentElement(el)) {
      return printArgumentElement(el);
    }
    if (isDateElement(el) || isTimeElement(el) || isNumberElement(el)) {
      return printSimpleFormatElement(el);
    }

    if (isPluralElement(el)) {
      return printPluralElement(el);
    }

    if (isSelectElement(el)) {
      return printSelectElement(el);
    }

    if (isPoundElement(el)) {
      return '#';
    }
  });

  return printedNodes.join('');
}

function printEscapedMessage(message: string): string {
  return message.replace(/([{}](?:.*[{}])?)/su, `'$1'`);
}

function printLiteralElement({value}: LiteralElement, isInPlural: boolean) {
  const escaped = printEscapedMessage(value);
  return isInPlural ? escaped.replace('#', "'#'") : escaped;
}

function printArgumentElement({value}: ArgumentElement) {
  return `{${value}}`;
}

function printSimpleFormatElement(
  el: DateElement | TimeElement | NumberElement
) {
  return `{${el.value}, ${TYPE[el.type]}${
    el.style ? `, ${printArgumentStyle(el.style)}` : ''
  }}`;
}

function printNumberSkeletonToken(token: NumberSkeletonToken): string {
  const {stem, options} = token;
  return options.length === 0
    ? stem
    : `${stem}${options.map(o => `/${o}`).join('')}`;
}

function printArgumentStyle(style: string | Skeleton) {
  if (typeof style === 'string') {
    return printEscapedMessage(style);
  } else if (style.type === SKELETON_TYPE.dateTime) {
    return `::${printDateTimeSkeleton(style)}`;
  } else {
    return `::${style.tokens.map(printNumberSkeletonToken).join(' ')}`;
  }
}

export function printDateTimeSkeleton(style: DateTimeSkeleton): string {
  return style.pattern;
}

function printSelectElement(el: SelectElement) {
  const msg = [
    el.value,
    'select',
    Object.keys(el.options)
      .map(id => `${id}{${doPrintAST(el.options[id].value, false)}}`)
      .join(' '),
  ].join(',');
  return `{${msg}}`;
}

function printPluralElement(el: PluralElement) {
  const type = el.pluralType === 'cardinal' ? 'plural' : 'selectordinal';
  const msg = [
    el.value,
    type,
    [
      el.offset ? `offset:${el.offset}` : '',
      ...Object.keys(el.options).map(
        id => `${id}{${doPrintAST(el.options[id].value, true)}}`
      ),
    ]
      .filter(Boolean)
      .join(' '),
  ].join(',');
  return `{${msg}}`;
}
