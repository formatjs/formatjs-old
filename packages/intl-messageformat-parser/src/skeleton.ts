import {NumberSkeletonToken} from './types';
import {UnifiedNumberFormatOptions} from '@formatjs/intl-unified-numberformat';

/**
 * https://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table
 * Credit: https://github.com/caridy/intl-datetimeformat-pattern/blob/master/index.js
 * with some tweaks
 */
const DATE_TIME_REGEX = /(?:[Eec]{1,6}|G{1,5}|[Qq]{1,5}|(?:[yYur]+|U{1,5})|[ML]{1,5}|d{1,2}|D{1,3}|F{1}|[abB]{1,5}|[hkHK]{1,2}|w{1,2}|W{1}|m{1,2}|s{1,2}|[zZOvVxX]{1,4})(?=([^']*'[^']*')*[^']*$)/g;

export interface ExtendedDateTimeFormatOptions
  extends Intl.DateTimeFormatOptions {
  hourCycle?: 'h11' | 'h12' | 'h23' | 'h24';
}

/**
 * Parse Date time skeleton into Intl.DateTimeFormatOptions
 * Ref: https://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table
 * @public
 * @param skeleton skeleton string
 */
export function parseDateTimeSkeleton(
  skeleton: string
): ExtendedDateTimeFormatOptions {
  const result: ExtendedDateTimeFormatOptions = {};
  skeleton.replace(DATE_TIME_REGEX, match => {
    const len = match.length;
    switch (match[0]) {
      // Era
      case 'G':
        result.era = len === 4 ? 'long' : len === 5 ? 'narrow' : 'short';
        break;
      // Year
      case 'y':
        result.year = len === 2 ? '2-digit' : 'numeric';
        break;
      case 'Y':
      case 'u':
      case 'U':
      case 'r':
        throw new RangeError(
          '`Y/u/U/r` (year) patterns are not supported, use `y` instead'
        );
      // Quarter
      case 'q':
      case 'Q':
        throw new RangeError('`q/Q` (quarter) patterns are not supported');
      // Month
      case 'M':
      case 'L':
        result.month = ['numeric', '2-digit', 'short', 'long', 'narrow'][
          len - 1
        ];
        break;
      // Week
      case 'w':
      case 'W':
        throw new RangeError('`w/W` (week) patterns are not supported');
      case 'd':
        result.day = ['numeric', '2-digit'][len - 1];
        break;
      case 'D':
      case 'F':
      case 'g':
        throw new RangeError(
          '`D/F/g` (day) patterns are not supported, use `d` instead'
        );
      // Weekday
      case 'E':
        result.weekday = len === 4 ? 'short' : len === 5 ? 'narrow' : 'short';
        break;
      case 'e':
        if (len < 4) {
          throw new RangeError('`e..eee` (weekday) patterns are not supported');
        }
        result.weekday = ['short', 'long', 'narrow', 'short'][len - 4];
        break;
      case 'c':
        if (len < 4) {
          throw new RangeError('`c..ccc` (weekday) patterns are not supported');
        }
        result.weekday = ['short', 'long', 'narrow', 'short'][len - 4];
        break;

      // Period
      case 'a': // AM, PM
        result.hour12 = true;
        break;
      case 'b': // am, pm, noon, midnight
      case 'B': // flexible day periods
        throw new RangeError(
          '`b/B` (period) patterns are not supported, use `a` instead'
        );
      // Hour
      case 'h':
        result.hourCycle = 'h12';
        result.hour = ['numeric', '2-digit'][len - 1];
        break;
      case 'H':
        result.hourCycle = 'h23';
        result.hour = ['numeric', '2-digit'][len - 1];
        break;
      case 'K':
        result.hourCycle = 'h11';
        result.hour = ['numeric', '2-digit'][len - 1];
        break;
      case 'k':
        result.hourCycle = 'h24';
        result.hour = ['numeric', '2-digit'][len - 1];
        break;
      case 'j':
      case 'J':
      case 'C':
        throw new RangeError(
          '`j/J/C` (hour) patterns are not supported, use `h/H/K/k` instead'
        );
      // Minute
      case 'm':
        result.minute = ['numeric', '2-digit'][len - 1];
        break;
      // Second
      case 's':
        result.second = ['numeric', '2-digit'][len - 1];
        break;
      case 'S':
      case 'A':
        throw new RangeError(
          '`S/A` (second) pattenrs are not supported, use `s` instead'
        );
      // Zone
      case 'z': // 1..3, 4: specific non-location format
        result.timeZoneName = len < 4 ? 'short' : 'long';
        break;
      case 'Z': // 1..3, 4, 5: The ISO8601 varios formats
      case 'O': // 1, 4: miliseconds in day short, long
      case 'v': // 1, 4: generic non-location format
      case 'V': // 1, 2, 3, 4: time zone ID or city
      case 'X': // 1, 2, 3, 4: The ISO8601 varios formats
      case 'x': // 1, 2, 3, 4: The ISO8601 varios formats
        throw new RangeError(
          '`Z/O/v/V/X/x` (timeZone) pattenrs are not supported, use `z` instead'
        );
    }
    return '';
  });
  return result;
}

function icuUnitToEcma(unit: string): UnifiedNumberFormatOptions['unit'] {
  return unit.replace(/^(.*?)-/, '') as UnifiedNumberFormatOptions['unit'];
}

const FRACTION_PRECISION_REGEX = /^\.(?:(0+)(\*)?|(#+)|(0+)(#+))$/g;
const SIGNIFICANT_PRECISION_REGEX = /^(@+)?(\+|#+)?$/g;

function parseSignificantPrecision(str: string): UnifiedNumberFormatOptions {
  const result: UnifiedNumberFormatOptions = {};
  str.replace(SIGNIFICANT_PRECISION_REGEX, function (
    _: string,
    g1: string,
    g2: string | number
  ) {
    // @@@ case
    if (typeof g2 !== 'string') {
      result.minimumSignificantDigits = g1.length;
      result.maximumSignificantDigits = g1.length;
    }
    // @@@+ case
    else if (g2 === '+') {
      result.minimumSignificantDigits = g1.length;
    }
    // .### case
    else if (g1[0] === '#') {
      result.maximumSignificantDigits = g1.length;
    }
    // .@@## or .@@@ case
    else {
      result.minimumSignificantDigits = g1.length;
      result.maximumSignificantDigits =
        g1.length + (typeof g2 === 'string' ? g2.length : 0);
    }
    return '';
  });
  return result;
}

function parseSign(str: string): UnifiedNumberFormatOptions | undefined {
  switch (str) {
    case 'sign-auto':
      return {
        signDisplay: 'auto',
      };
    case 'sign-accounting':
      return {
        currencySign: 'accounting',
      };
    case 'sign-always':
      return {
        signDisplay: 'always',
      };
    case 'sign-accounting-always':
      return {
        signDisplay: 'always',
        currencySign: 'accounting',
      };
    case 'sign-except-zero':
      return {
        signDisplay: 'exceptZero',
      };
    case 'sign-accounting-except-zero':
      return {
        signDisplay: 'exceptZero',
        currencySign: 'accounting',
      };
    case 'sign-never':
      return {
        signDisplay: 'never',
      };
  }
}

function parseNotationOptions(opt: string): UnifiedNumberFormatOptions {
  const result: UnifiedNumberFormatOptions = {};
  const signOpts = parseSign(opt);
  if (signOpts) {
    return signOpts;
  }
  return result;
}

/**
 * https://github.com/unicode-org/icu/blob/master/docs/userguide/format_parse/numbers/skeletons.md#skeleton-stems-and-options
 */
export function convertNumberSkeletonToNumberFormatOptions(
  tokens: NumberSkeletonToken[]
): UnifiedNumberFormatOptions {
  let result: UnifiedNumberFormatOptions = {};
  for (const token of tokens) {
    switch (token.stem) {
      case 'percent':
        result.style = 'percent';
        continue;
      case 'currency':
        result.style = 'currency';
        result.currency = token.options[0];
        continue;
      case 'group-off':
        result.useGrouping = false;
        continue;
      case 'precision-integer':
      case '.':
        result.maximumFractionDigits = 0;
        continue;
      case 'measure-unit':
        result.style = 'unit';
        result.unit = icuUnitToEcma(token.options[0]);
        continue;
      case 'compact-short':
        result.notation = 'compact';
        result.compactDisplay = 'short';
        continue;
      case 'compact-long':
        result.notation = 'compact';
        result.compactDisplay = 'long';
        continue;
      case 'scientific':
        result = {
          ...result,
          notation: 'scientific',
          ...token.options.reduce(
            (all, opt) => ({...all, ...parseNotationOptions(opt)}),
            {}
          ),
        };
        continue;
      case 'engineering':
        result = {
          ...result,
          notation: 'engineering',
          ...token.options.reduce(
            (all, opt) => ({...all, ...parseNotationOptions(opt)}),
            {}
          ),
        };
        continue;
      case 'notation-simple':
        result.notation = 'standard';
        continue;
      // https://github.com/unicode-org/icu/blob/master/icu4c/source/i18n/unicode/unumberformatter.h
      case 'unit-width-narrow':
        result.currencyDisplay = 'narrowSymbol';
        result.unitDisplay = 'narrow';
        continue;
      case 'unit-width-short':
        result.currencyDisplay = 'code';
        result.unitDisplay = 'short';
        continue;
      case 'unit-width-full-name':
        result.currencyDisplay = 'name';
        result.unitDisplay = 'long';
        continue;
      case 'unit-width-iso-code':
        result.currencyDisplay = 'symbol';
        continue;
    }
    // Precision
    // https://github.com/unicode-org/icu/blob/master/docs/userguide/format_parse/numbers/skeletons.md#fraction-precision
    // precision-integer case
    if (FRACTION_PRECISION_REGEX.test(token.stem)) {
      if (token.options.length > 1) {
        throw new RangeError(
          'Fraction-precision stems only accept a single optional option'
        );
      }
      token.stem.replace(FRACTION_PRECISION_REGEX, function (
        match: string,
        g1: string,
        g2: string | number,
        g3: string,
        g4: string,
        g5: string
      ) {
        // .000* case (before ICU67 it was .000+)
        if (g2 === '*') {
          result.minimumFractionDigits = g1.length;
        }
        // .### case
        else if (g3 && g3[0] === '#') {
          result.maximumFractionDigits = g3.length;
        }
        // .00## case
        else if (g4 && g5) {
          result.minimumFractionDigits = g4.length;
          result.maximumFractionDigits = g4.length + g5.length;
        } else {
          result.minimumFractionDigits = g1.length;
          result.maximumFractionDigits = g1.length;
        }
        return '';
      });

      if (token.options.length) {
        result = {...result, ...parseSignificantPrecision(token.options[0])};
      }
      continue;
    }
    if (SIGNIFICANT_PRECISION_REGEX.test(token.stem)) {
      result = {...result, ...parseSignificantPrecision(token.stem)};
      continue;
    }
    const signOpts = parseSign(token.stem);
    if (signOpts) {
      result = {...result, ...signOpts};
    }
  }
  return result;
}
