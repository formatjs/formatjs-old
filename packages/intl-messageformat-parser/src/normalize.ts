import {
  MessageFormatElement,
  isPluralElement,
  isLiteralElement,
  LiteralElement
} from './types';
import { parse } from './parser';

const PLURAL_HASHTAG_REGEX = /(^|[^\\])#/g;

/**
 * Whether to convert `#` in plural rule options
 * to `{var, number}`
 * @param el AST Element
 * @param pluralStack current plural stack
 */
export function normalizeHashtagInPlural(els: MessageFormatElement[]) {
  els.forEach(el => {
    // If we're encountering a plural el
    if (!isPluralElement(el)) {
      return;
    }
    // Go down the options and search for # in any literal element
    Object.keys(el.options).forEach(id => {
      const opt = el.options[id];
      // If we got a match, we have to split this
      // and inject a NumberElement in the middle
      let matchingLiteralElIndex = -1;
      let literalEl: LiteralElement | undefined = undefined;
      for (let i = 0; i < opt.value.length; i++) {
        const el = opt.value[i] as LiteralElement;
        if (isLiteralElement(el) && PLURAL_HASHTAG_REGEX.test(el.value)) {
          matchingLiteralElIndex = i;
          literalEl = el;
          break;
        }
      }

      if (literalEl) {
        const newValue = literalEl.value.replace(
          PLURAL_HASHTAG_REGEX,
          `$1{${el.value}, number}`
        );
        const newEls = parse(newValue);
        opt.value.splice(matchingLiteralElIndex, 1, ...newEls);
      }
      normalizeHashtagInPlural(opt.value);
    });
  });
}
