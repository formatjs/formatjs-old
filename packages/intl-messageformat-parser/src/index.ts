import { parse as pegParse, IParseOptions } from './parser';
import { Options, MessageFormatElement } from './types';
import { normalizeHashtagInPlural } from './normalize';
export * from './types';
export * from './parser';

export type ParseOptions = Options & IParseOptions;

export function parse(
  input: string,
  opts?: ParseOptions
): MessageFormatElement[] {
  const els = pegParse(input, opts);
  if (!opts || opts.normalizeHashtagInPlural !== false) {
    normalizeHashtagInPlural(els);
  }
  return els;
}
