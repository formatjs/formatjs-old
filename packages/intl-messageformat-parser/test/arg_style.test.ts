import { parse } from '../src/parser';
import { printAST } from '../src/printer';

test.each([
  `{0,date,y-M-d HH:mm:ss}`,
  `{0,number,y-M-d HH:mm:ss}`,
  `{0,date,y-M-d HH:mm:ss zzzz}`,
  `{0,number,y-M-d HH:mm:ss zzzz}`,
  `{0,date,y-M-d,HH:mm:ss,zzzz}`,
  `{0,number,y-M-d,HH:mm:ss,zzzz}`,
  `{0,number,'{}'}`,
  `{0,number,''}`
])('argStyleText test case: %p', testCase => {
  const ast = parse(testCase);
  expect(ast).toMatchSnapshot();
  expect(printAST(ast)).toMatchSnapshot();
});
