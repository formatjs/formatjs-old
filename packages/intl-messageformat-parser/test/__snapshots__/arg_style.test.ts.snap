// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`argStyleText test case: "{0,date,y-M-d HH:mm:ss zzzz}" 1`] = `
Array [
  Object {
    "style": "y-M-d HH:mm:ss zzzz",
    "type": 3,
    "value": "0",
  },
]
`;

exports[`argStyleText test case: "{0,date,y-M-d HH:mm:ss zzzz}" 2`] = `"{0, date, y-M-d HH:mm:ss zzzz}"`;

exports[`argStyleText test case: "{0,date,y-M-d HH:mm:ss}" 1`] = `
Array [
  Object {
    "style": "y-M-d HH:mm:ss",
    "type": 3,
    "value": "0",
  },
]
`;

exports[`argStyleText test case: "{0,date,y-M-d HH:mm:ss}" 2`] = `"{0, date, y-M-d HH:mm:ss}"`;

exports[`argStyleText test case: "{0,date,y-M-d,HH:mm:ss,zzzz}" 1`] = `
Array [
  Object {
    "style": "y-M-d,HH:mm:ss,zzzz",
    "type": 3,
    "value": "0",
  },
]
`;

exports[`argStyleText test case: "{0,date,y-M-d,HH:mm:ss,zzzz}" 2`] = `"{0, date, y-M-d,HH:mm:ss,zzzz}"`;

exports[`argStyleText test case: "{0,number,''}" 1`] = `
Array [
  Object {
    "style": "'",
    "type": 2,
    "value": "0",
  },
]
`;

exports[`argStyleText test case: "{0,number,''}" 2`] = `"{0, number, '}"`;

exports[`argStyleText test case: "{0,number,'{}'}" 1`] = `
Array [
  Object {
    "style": "{}",
    "type": 2,
    "value": "0",
  },
]
`;

exports[`argStyleText test case: "{0,number,'{}'}" 2`] = `"{0, number, '{}'}"`;

exports[`argStyleText test case: "{0,number,y-M-d HH:mm:ss zzzz}" 1`] = `
Array [
  Object {
    "style": "y-M-d HH:mm:ss zzzz",
    "type": 2,
    "value": "0",
  },
]
`;

exports[`argStyleText test case: "{0,number,y-M-d HH:mm:ss zzzz}" 2`] = `"{0, number, y-M-d HH:mm:ss zzzz}"`;

exports[`argStyleText test case: "{0,number,y-M-d HH:mm:ss}" 1`] = `
Array [
  Object {
    "style": "y-M-d HH:mm:ss",
    "type": 2,
    "value": "0",
  },
]
`;

exports[`argStyleText test case: "{0,number,y-M-d HH:mm:ss}" 2`] = `"{0, number, y-M-d HH:mm:ss}"`;

exports[`argStyleText test case: "{0,number,y-M-d,HH:mm:ss,zzzz}" 1`] = `
Array [
  Object {
    "style": "y-M-d,HH:mm:ss,zzzz",
    "type": 2,
    "value": "0",
  },
]
`;

exports[`argStyleText test case: "{0,number,y-M-d,HH:mm:ss,zzzz}" 2`] = `"{0, number, y-M-d,HH:mm:ss,zzzz}"`;
